File: core/LengthOfLongestFibonacciSubsequence/LengthOfLongestFibonacciSubsequence.py
from typing import List


def lenLongestFibSubseq(arr: List[int]) -> int:
    res, length = 0, len(arr)
    indices = {num: index for index, num in enumerate(arr)}
    dp = [[0] * length for _ in range(length)]
    for i, x in enumerate(arr):
        for j in range(i - 1, -1, -1):
            if arr[j] * 2 <= x:
                break
            if x - arr[j] in indices:
                k = indices[x - arr[j]]
                dp[j][i] = max(dp[k][j] + 1, 3)
                res = max(res, dp[j][i])

    return res

File: core/LengthOfLongestFibonacciSubsequence/LengthOfLongestFibonacciSubsequence.java
package com.dbc.code;

import java.util.HashMap;
import java.util.Map;

public class LengthOfLongestFibonacciSubsequence {
    public int lenLongestFibSubseq(int[] arr) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < arr.length; i++) map.put(arr[i], i);
        int[][] dp = new int[arr.length][arr.length];
        int res = 0;
        for (int i = 0; i < arr.length; i++) {
            for (int j = i - 1; j >= 0; j--) {
                if (arr[j] * 2 <= arr[i]) break;
                if (map.containsKey(arr[i] - arr[j])) {
                    int k = map.get(arr[i] - arr[j]);
                    dp[j][i] = Math.max(dp[k][j] + 1, 3);
                    res = Math.max(res, dp[j][i]);
                }
            }
        }
        return res;
    }
}
